llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp

```diff
@@ -1074,6 +1074,35 @@ void AsmPrinter::emitFunctionHeader() {
 
 static bool needFuncLabelsForEH(const MachineFunction &MF);
 
+/// Returns true if function begin label should be emitted due to a
+/// BlockAddress used as an initializer for a GlobalVariable.
+static bool needFuncLabelForBlockAddress(const MachineFunction &MF) {
+  SmallVector<const Value *, 4> Users;
+
+  for (const auto &MBB : MF) {
+    if (!MBB.isIRBlockAddressTaken())
+      continue;
+
+    const auto *BB = MBB.getAddressTakenIRBlock();
+    for (const auto &U : BB->users()) {
+      if (isa<BlockAddress>(U))
+        Users.push_back(U);
+    }
+  }
+
+  while (!Users.empty()) {
+    const Value *V = Users.pop_back_val();
+    if (isa<GlobalVariable>(V))
```

Comment:
Do we actually need this extra logic? Wouldn't it be fine to always emit the start label if we see a blockaddress use? Worst case is that we added an unnecessary symtab entry to the .o file.

I'm not opposed to this extra checking to avoid redundant labels just thinking it would be simpler and safer in case of some future unknown IR change. What do you think?

Reply:
So, this pattern is actually exceedingly rare. In part it's because I don't want any test using BlockAddress for something to have the label, because anything that's not a global doesn't need this. I'd prefer not to emit labels when we can relatively easily determine they're not needed, so I'd rather go with this for now and if it becomes brittle we can reconsider using a more conservative overapproximation?